File: leetcode/reverseList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  if (!head.next || !head) return head;
  let p = reverseList(head.next);
  head.next.next = head;
  head.next = null;
  return p;
};

var reverseList = function (head) {
  let prev = null;
  let cur = head;
  while (cur !== null) {
    let next = cur.next;
    cur.next = prev;
    prev = cur;
    cur = next;
  }
  return prev;
};
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}
const head = new ListNode(
  1,
  new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5))))
);
reverseList(head);

File: leetcode/reverseList.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(
            self, head: Optional[ListNode]) -> Optional[ListNode]:
        return reverse_list(head)


def reverse_list(node, pre=None):
    if node is None:
        return pre

    nxt = node.next
    node.next = pre

    return reverse_list(nxt, node)


def reverse_list_iterative(head):
    prev = None
    cur = head
    while cur:
        cur.next, prev, cur = prev, cur, cur.next
    return prev


list1 = ListNode(1, ListNode(
    2, ListNode(3, ListNode(4, ListNode(5)))))

a = reverse_list(list1)

while a:
    print(a.val)
    a = a.next
